File: iMessageExplorer/database_helper.py
import sqlite3
import json
import sys

def fetch_participants(db_path):
    try:
        connection = sqlite3.connect(db_path)
        cursor = connection.cursor()
        
        cursor.execute("SELECT DISTINCT id FROM handle")
        rows = cursor.fetchall()
        participants = [row[0] for row in rows if row[0] is not None]
        
        print(json.dumps(participants), flush=True)
    except sqlite3.Error as e:
        print(json.dumps({"error": str(e)}), flush=True)
    finally:
        if connection:
            connection.close()

def fetch_messages(db_path, participant):
    try:
        connection = sqlite3.connect(db_path)
        cursor = connection.cursor()
        
        cursor.execute("""
            SELECT message.text 
            FROM message
            JOIN handle ON message.handle_id = handle.ROWID
            WHERE handle.id = ? AND message.text IS NOT NULL
        """, (participant,))
        rows = cursor.fetchall()
        messages = [row[0] for row in rows]
        
        print(json.dumps(messages), flush=True)
    except sqlite3.Error as e:
        print(json.dumps({"error": str(e)}), flush=True)
    finally:
        if connection:
            connection.close()

if __name__ == "__main__":
    db_path = sys.argv[1]
    action = sys.argv[2]
    
    if action == "participants":
        fetch_participants(db_path)
    elif action == "messages" and len(sys.argv) > 3:
        participant = sys.argv[3]
        fetch_messages(db_path, participant)
    else:
        print(json.dumps({"error": "Invalid arguments"}), flush=True)

File: iMessageExplorer/DatabaseHelper.swift
//
//  DatabaseHelper.swift
//  iMessageExplorer
//
//  Created by Raqeeb Anjum on 12/14/24.
//

import SQLite
import Foundation

struct DatabaseHelper {
    static func fetchMessages(from folder: URL) -> [Message] {
        var messages = [Message]()
        
        let dbPath = folder.appendingPathComponent("chat.db").path
        
        guard FileManager.default.fileExists(atPath: dbPath) else {
            print("Database file does not exist at path: \(dbPath)")
            return messages
        }
        
        do {
            let db = try Connection(dbPath, readonly: true)
            
            #if DEBUG
            db.trace { print($0) }
            #endif
            
            let messageTable = Table("message")
            let id = Expression<Int64>("ROWID")
            let text = Expression<String?>("text")
            let isFromMe = Expression<Int64>("is_from_me")
            let date = Expression<Double>("date")
            
            // List available tables
            print("Checking available tables:")
            let tableNames = try db.scalar("SELECT name FROM sqlite_master WHERE type='table';") as? [String] ?? []
            for tableName in tableNames {
                print("Found table: \(tableName)")
            }
            
            let query = messageTable.order(date.asc)
            
            let rowCount = try db.scalar(query.count)
            print("Total rows in message table: \(rowCount)")
            
            let iterator = try db.prepare(query)
            
            for row in iterator {
                print("Raw row data:")
                print("ID: \(row[id])")
                print("Text: \(row[text] ?? "nil")")
                print("Is From Me: \(row[isFromMe])")
                print("Date: \(row[date])")

                let messageText = row[text] ?? "(No text)"
                let timestamp = Date(timeIntervalSince1970: row[date] + 978307200)
                let message = Message(
                    id: Int(row[id]),
                    text: messageText,
                    timestamp: timestamp,
                    isFromMe: row[isFromMe] == 1
                )
                messages.append(message)
            }
        } catch {
            print("Error details:")
            print("Error type: \(type(of: error))")
            print("Error description: \(error.localizedDescription)")
            print("Error: \(error)")
        }
        
        return messages
    }
}
